worlds/pokemon_emerald/opponents.py

```diff
@@ -0,0 +1,116 @@
+    }
+    should_match_type = world.options.trainer_parties in {
+        RandomizeTrainerParties.option_match_type,
+        RandomizeTrainerParties.option_match_base_stats_and_type,
+    }
+
+    per_species_tmhm_moves: Dict[int, List[int]] = {}
+
+    for trainer in world.modified_trainers:
+        new_party = []
+        for pokemon in trainer.party.pokemon:
+            original_species = data.species[pokemon.species_id]
+
+            # Construct progressive tiers of blacklists that can be peeled back if they
+            # collectively cover too much of the pokedex. A lower index in `blacklists`
+            # indicates a more important set of species to avoid. Entries at `0` will
+            # always be blacklisted.
+            blacklists: Dict[int, List[Set[int]]] = defaultdict(list)
+
+            # Blacklist unevolved species
+            if pokemon.level >= world.options.force_fully_evolved:
+                blacklists[0].append(UNEVOLVED_POKEMON)
+
+            # Blacklist from player options
+            blacklists[2].append(world.blacklisted_opponent_pokemon)
+
+            # Type matching blacklist
+            if should_match_type:
+                blacklists[3].append({
+                    species.species_id
+                    for species in world.modified_species.values()
+                    if not bool(set(species.types) & set(original_species.types))
+                })
+
+            merged_blacklist: Set[int] = set()
+            for max_priority in reversed(sorted(blacklists.keys())):
+                merged_blacklist = set()
+                for priority in blacklists.keys():
+                    if priority <= max_priority:
+                        for blacklist in blacklists[priority]:
+                            merged_blacklist |= blacklist
+
+                if len(merged_blacklist) < NUM_REAL_SPECIES:
+                    break
+            else:
+                raise RuntimeError("This should never happen")
+
+            candidates = [
+                species
+                for species in world.modified_species.values()
+                if species.species_id not in merged_blacklist
+            ]
+
+            if should_match_bst:
+                candidates = filter_species_by_nearby_bst(candidates, sum(original_species.base_stats))
+
+            new_species = world.random.choice(candidates)
+
+            if new_species.species_id not in per_species_tmhm_moves:
+                per_species_tmhm_moves[new_species.species_id] = list({
+                    world.modified_tmhm_moves[i]
+                    for i, is_compatible in enumerate(int_to_bool_array(new_species.tm_hm_compatibility))
+                    if is_compatible
+                })
+
+            # TMs and HMs compatible with the species
+            tm_hm_movepool = per_species_tmhm_moves[new_species.species_id]
+
+            # Moves the pokemon could have learned by now
+            level_up_movepool = list({
```

Comment:
what's the reasoning for this? it's nondeterministic.

Reply:
Good point. The purpose is to remove duplicates as I go. Switched to `sorted`